packages/core/contracts/financial-templates/funding-rate-store/implementation/OptimisticOracle.sol

```diff
@@ -0,0 +1,357 @@
+        require(getState(msg.sender, identifier, timestamp) == State.Requested, "setRefundOnDispute: Requested");
+        request.refundOnDispute = true;
+    }
+
+    function setCustomLiveness(
+        bytes32 identifier,
+        uint256 timestamp,
+        uint256 customLiveness
+    ) external nonReentrant() {
+        Request storage request = _getRequest(msg.sender, identifier, timestamp);
+        _validateLiveness(customLiveness);
+        require(getState(msg.sender, identifier, timestamp) == State.Requested, "setCustomLiveness: Requested");
+        request.customLiveness = customLiveness;
+    }
+
+    function proposePriceFor(
+        address proposer,
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        int256 proposedPrice
+    ) public nonReentrant() {
+        Request storage request = _getRequest(requester, identifier, timestamp);
+        require(getState(requester, identifier, timestamp) == State.Requested, "proposePriceFor: Requested");
+        request.proposer = proposer;
+        request.proposedPrice = proposedPrice;
+        // If a custom liveness has been set
+        request.expirationTime = getCurrentTime().add(
+            request.customLiveness != 0 ? request.customLiveness : defaultLiveness
+        );
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond.add(request.finalFee));
+
+        // Event.
+        emit ProposePrice(requester, proposer, identifier, timestamp, proposedPrice);
+
+        // Callback.
+        try FundingRateRequester(requester).priceProposed(identifier, timestamp) {} catch {}
+    }
+
+    function proposePrice(
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        int256 proposedPrice
+    ) external {
+        // Note: re-entrancy guard is done in the inner call.
+        proposePriceFor(msg.sender, requester, identifier, timestamp, proposedPrice);
+    }
+
+    function disputePriceFor(
+        address disputer,
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp
+    ) public nonReentrant() {
+        Request storage request = _getRequest(requester, identifier, timestamp);
+        require(getState(requester, identifier, timestamp) == State.Proposed, "disputePriceFor: Proposed");
+        request.disputer = disputer;
+        request.currency.safeTransferFrom(msg.sender, address(this), request.bond.add(request.finalFee));
+        _getOracle().requestPrice(identifier, timestamp);
+        StoreInterface store = _getStore();
+        request.currency.safeIncreaseAllowance(address(store), request.finalFee);
+        _getStore().payOracleFeesErc20(address(request.currency), FixedPoint.Unsigned(request.finalFee));
+
+        // Compute refund.
+        uint256 refund = 0;
+        if (request.reward != 0 && request.refundOnDispute) {
+            refund = request.reward;
```

Comment:
do you not need to transfer this refund in the `_settle` function or send it back here?

Reply:
Ha, oops! Great catch.

Reply:
Done.